testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/SecureRedirectUrisEnforcerExecutorTest.java

```diff
@@ -180,7 +182,7 @@ public void testSecureRedirectUrisEnforcerExecutor_IPv6LoopbackAddress() throws
 
         // register - fail
         // IPv4 loopback address not allowed
-        testSecureRedirectUrisEnforcerExecutor_failRegisterDynamically(Arrays.asList("https://[::1]/", "https://localhost/auth/admin"));
+        testSecureRedirectUrisEnforcerExecutor_failRegisterDynamically(Arrays.asList("https://[::1]/", "https://127.0.0.1/auth/admin"));
```

Comment:
Question 1: Request does not fail when `localhost` is used (request was not failing before my changes as well, but test did not correctly tested that request fails).

In other words, it means that redirect-uri like `http://localhost/foo` is allowed even if configuration has `Allow IPv4 loopback` and `Allow IPv6 loopback` switched to false. This does not look like correct behaviour? WDYT?

I figured the issue is in guava code:

```
            try {
                addr = InetAddresses.forUriString(uri.getHost());
            } catch (IllegalArgumentException e) {
                return false;
            }
```
In case of `uri.getHost()` is `localhost`, there is IllegalArgumentException thrown and method returns false considering `localhost` as not being loopback address.

I wonder if we can use `InetAddress.getByName(uri.getHost())` instead of `InetAddresses.forUriString(uri.getHost())` ? This is the same, which is used for instance by `SslRequired` checks in Keycloak. Does it work for you?

Reply:
@mposolda 

> I wonder if we can use InetAddress.getByName(uri.getHost()) instead of InetAddresses.forUriString(uri.getHost()) ?

In the discussion in [the original PR](https://github.com/keycloak/keycloak/pull/25647), `InetAddresses.forUriString(uri.getHost())` instead of `InetAddress.getByName(uri.getHost()) ` to avoid a DNS lookup .

Reply:
@mposolda 

> In other words, it means that redirect-uri like http://localhost/foo is allowed even if configuration has Allow IPv4 loopback and Allow IPv6 loopback switched to false. This does not look like correct behaviour? WDYT?

+1 if `localhost` is not allowed if `oauth-2-1-compliant` is ON because OAuth 2.1 does not allow using `localhost`.

Reply:
@mposolda  I expect that registering a client dynamically with  redirect urls ("https://[::1]/", "https://127.0.0.1/auth/admin") fails because "Allow IPv4 loopback address" is OFF. Is that right?

Reply:
@tnorimat Yes, exactly. Registration of ("https://127.0.0.1/auth/admin") fails because `Allow IPv4 loopback address` is OFF. So when `127.0.0.1` is used in the address, it works as expected IMO as it is supposed to fail and it really fails.

But my point here is the other case when you try to register with use `localhost` instead of `127.0.0.1`. I think that registering of ("https://localhost/auth/admin") should fail in case of `Allow IPv4 loopback address` is `OFF`. But it currently does not fail. The client is successfully registered, which does not look like correct behaviour.

My observation is that `InetAddresses.forUriString(uri.getHost())` does not work well as in case when you call `InetAddresses.forUriString("localhost")` it throws `IllegalArgumentException`. This causes `SecureRedirectUrisEnforcerExecutor.isLoopbackAddressRedirectUri` to incorrectly return `false` even if `localhost` is loopback address.

ATM my vote is to add back `InetAddress.getByName` instead of `InetAddresses.forUriString` . I understand the concern with DNS lookup, but I am not sure if this is really big concern? For instance, Keycloak uses this already in `SslRequired.isLocal(String remoteAddress)` , which is called almost at every HTTP request to Keycloak?

Reply:
@mposolda Thank you for the clarification.

> I think that registering of ("https://localhost/auth/admin") should fail in case of Allow IPv4 loopback address is OFF. But it currently does not fail. The client is successfully registered, which does not look like correct behaviour.

Yes, I agree with you. We need to modify the executor to do so.

> ATM my vote is to add back InetAddress.getByName instead of InetAddresses.forUriString . I understand the concern with DNS lookup, but I am not sure if this is really big concern? For instance, Keycloak uses this already in SslRequired.isLocal(String remoteAddress) , which is called almost at every HTTP request to Keycloak?

@rmartinc suggested using `InetAddresses.forUriString`. I am OK to use `InetAddress.getByName`, but it might be better to ask @rmartinc for the point.

Reply:
@mposolda @tnorimat I recommended to avoid the DNS lookup because I saw the tests were OK. I indeed thought that localhost was not considered like loopback access because it is not recommended by the spec. If you thin it's OK just change for getByName again.